### ISSUE TITLE
FR: Pop-up on <textarea> controls

### ISSUE THREAD

**Issue by manyfew** [Paperfeed/LiuChan#7]
In Firefox I used add-on _Mandaring Popup_. I created a very simple website, stored locally on my hard drive as a text file, which displayed a number of multi-line text input boxes, here is the code:

```
<body>
	<pre> <textarea cols="50" rows="10"></textarea>  <textarea cols="51" rows="10"></textarea>  <textarea cols="50" rows="10"></textarea></pre>
	<pre> <textarea cols="50" rows="10"></textarea>  <textarea cols="51" rows="10"></textarea>  <textarea cols="50" rows="10"></textarea></pre>
	<pre> <textarea cols="50" rows="10"></textarea>  <textarea cols="51" rows="10"></textarea>  <textarea cols="50" rows="10"></textarea></pre>
</body>
```

I've been using this for over 7 years as a Chinese text reader, where I would paste, or type, a bunch of Chinese text and then casually hover the mouse over the _hanzi_ to get the English translation. Unfortunately, LiuChan doesn't appear to recognize text in `<textarea>` controls. It does, however, work on `<input>` controls. The difference is that the latter is a single-line input box, which is not very good for large texts:

> The `<textarea>` tag defines a multi-line text input control.

Would it be possible to enable LiuChan to work on `<textarea>` controls too? Thanks!!

**Comment by Paperfeed** [comment]
Hey Manuel!

That should definitely be possible. I'll have a look.

I've been working on your submitted issues but I'm a bit under the weather
atm so I'm not quite focused when I'm looking at my code. Progress is being
made however as I've put in most of the theme customization options today,
just need to fix a few bugs and improve usability before I release the
update. I think I'll try to fix all of the issues you've put up here before
I release though. Perhaps you'd be willing to do a beta test when it's
ready?

Sincerely,
Aldert

**Comment by manyfew** [comment]
With pleasure! And get well soon

**Comment by manyfew** [comment]
Interestingly, LiuChan does work on some `textarea` controls. For example, this one:

[https://www.chineseconverter.com/en/convert/chinese-to-pinyin](url)

Or even this one here on Github where I'm typing this right now. Although for some reason it suddenly decided to stop working. Cycling LiuChan off and on again didn't work, but after reloading the webpage it was working again. I noticed this behaviour with Perapera in the past, too.

I have inspected the page source to see if there was something I could change in my code to make it work, couldn't figure it out though, but obviously there's a way to do it.

**Comment by Paperfeed** [comment]
Yeah, this should be working, but I'm going to rewrite the entire highlighting engine, since it's an unreadable mess. Most of it is code from the original Rikaichan addon. 

If you notice any areas where it doesn't work, please send me the link where it won't work. Makes it much easier to debug! Cheers

**Comment by manyfew** [comment]
Hey, thanks for this update, so many nice features! I've just tested this on my textarea-based input box but it seems it doesn't work for me. I can't send you a link because it's just a local htm file, but you can copy the code into a text file and open it in Chrome:

```
<body>
	<pre> <textarea cols="70" rows="5"></textarea>  <textarea style=vocabulary cols="23" rows="5"></textarea></textarea></pre>
	<pre> <textarea cols="70" rows="5"></textarea>  <textarea cols="23" rows="5"></textarea></textarea></pre>
	<pre> <textarea cols="70" rows="5"></textarea>  <textarea cols="23" rows="5"></textarea></textarea></pre>
</body>


<style type="text/css">
	textarea {
		font-family: Microsoft Yahei, Geneva, sans-serif;
		font-size: 20pt;
		_resize: vertical;
		min-height: 2em;
		min-width: 2em;
	}
</style>
```

Save the file and type the path to the file path into Chrome's address bar, hit Enter and then type some Chinese characters in any of the  textarea boxes.

**Comment by Paperfeed** [comment]
When I did that last time I was unable to reproduce the bug, I was able to
'select'  text in those areas normally. I'll have another look later.

Like I said though I'll be rewriting the entire highlighting engine. It's a
bit of a mess and there's a bunch of weird bugs related to it.

**Comment by manyfew** [comment]
Cool, I look forward to that

**Comment by manyfew** [comment]
Is this well underway? :)
